rm pkg/workloads

aanm opened the pull request:
As we have deprecated the usage of any workload in v1.6 it's time to
remove the pkg/workloads entirely. The option flags will still be
available until v1.8.

Only one option was removed entirely which was --docker (or -e) as it
has been deprecated since v1.0.4.

Also, due this changes, besides the LOC removed, the cilium-agent binary
also decreased from 99MB -> 71M

Signed-off-by: André Martins <andre@cilium.io>

```release-note
Remove container runtime dependencies
```

Read per commit basis!

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9447)
<!-- Reviewable:end -->

aanm committed:
rm pkg/workloads

As we have deprecated the usage of any workload in v1.6 it's time to
remove the pkg/workloads entirely. The option flags will still be
available until v1.8.

Only one option was removed entirely which was --docker (or -e) as it
has been deprecated since v1.0.4.

Also, due this changes, besides the LOC removed the cilium-agent binary
also decreased from 99MB -> 71M

Signed-off-by: André Martins <andre@cilium.io>

brb reviewed (changes requested):
This breaks ipvlan in cilium-docker (the libnetwork plugin). We should probably just remove the support for ipvlan from `plugins/cilium-docker/driver/driver.go`. /cc @borkmann

ianvernon commented on Documentation/install/upgrade.rst:
```diff
@@ -275,12 +275,21 @@ IMPORTANT: Changes required before upgrading to 1.7.0
     key-file: '/var/lib/cilium/etcd-client.key'
     cert-file: '/var/lib/cilium/etcd-client.crt'
 
+* Due the removal of external libraries to connect to container runtimes, Cilium
```
"Due to the"

ianvernon commented on Documentation/install/upgrade.rst:
```diff
@@ -275,12 +275,21 @@ IMPORTANT: Changes required before upgrading to 1.7.0
   a particular device, consider using :ref:`nodeport`.
 
+* ``docker`` and ``e``: This flags has been removed as Cilium no longer requires
+  container runtimes integrations to manage containers' networks.
```
container runtime integrations

ianvernon commented on install/kubernetes/quick-install.yaml:
```diff
@@ -463,6 +454,7 @@ spec:
         - mountPath: /var/run/cilium
           name: cilium-run
       restartPolicy: Always
+      priorityClassName: system-node-critical
```
I don't think this should be here - I recall that this got added when running `make` in the `install/kubernetes` directory when I was releasing a micro-release of `v1.6` and @tgraf said it shouldn't be there - @tgraf can you confirm?

ianvernon reviewed (commented):
How does this work in the context of our `libnetwork` plugin overall? Is the only main impact the loss of the managing of existing container with Flannel?

aanm committed:
daemon: add better debug messages to log {put,patch}EndpointID

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
runtime: remove RuntimeConnectivityInVethModeTest

Since we are going to remove the container event watcher from the daemon
we will not be able to get the labels of containers started without
kubernetes and running without libnetwork. In other words, we will not
be able to get the container labels of containers for which they will
have their network created with CNI and are running outside of a k8s
environment.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
plugns/cilium-docker: add docker client to docker plugin

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/labels check if the map is zero instead of nil

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/endpoint: do not run duplicate UpdateLogger calls

UpdateLogger is already executed inside SetDefaultOpts

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/endpoint.EndpointChange: accept labels, container name and container ID

In order to modify an already running endpoint, cilium-docker plugin
will need to modify these fields from the endpoint structure. This is
because cilium-docker is the entity in charge to watch for docker
events and therefore will be able to know which labels are assigned for
each container started in a docker. Once it receive a new "start" event,
it will send a EndpointChange to cilium-agent so that cilium-agent is
aware of the labels associated with the recently endpoint created.

from endpoint request changes

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/endpoint: allow removing the endpoint without changing labels

Since cilium-agent is no longer able to detect the labels of a docker
container because it does not have any container watcher mechanism, it
needs to rely on cilium-docker plugin to remove the endpoints from the
init state. Once cilium-docker plugin sees the "start" event for the
recently created container, it will send a "modify labels" request with
the labels received in the "start" event which allows cilium-agent to
remove that endpoint from its init state. As not all containers can have
labels associated with them, cilium-agent needs to be more tolerant
and allow "modify labels" requests without any label in that request.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
cilium-docker/driver: add event watcher in cilium-docker plugin

Since pkg/workloads was removed from cilium-agent, all code dedicated
for docker will reside in `plugins/cilium-docker` including a event
watcher for docker events which allows to inform cilium-agent the labels
of a new initialized container.

Signed-off-by: André Martins <andre@cilium.io>

ianvernon commented on Documentation/install/upgrade.rst:
```diff
@@ -277,12 +277,21 @@ IMPORTANT: Changes required before upgrading to 1.7.0
 
+* Due to the removal of external libraries to connect to container runtimes,
+  Cilium no longer supports the option ``flannel-manage-existing-containers``.
+  Cilium will still support integration with flannel for new containers
```
casing: flannel --> Flannel

ianvernon commented on pkg/option/config.go:
```diff
@@ -110,9 +110,11 @@ const (
 	// ContainerRuntime sets the container runtime(s) used by Cilium
 	// { containerd | crio | docker | none | auto } ( "auto" uses the container
 	// runtime found in the order: "docker", "containerd", "crio" )
+	// Deprecated
```
Make this `Deprecated: <description>` c.f.: https://github.com/golang/go/wiki/Deprecated

ianvernon commented on pkg/option/config.go:
```diff
@@ -110,9 +110,11 @@ const (
 	ContainerRuntime = "container-runtime"
 
 	// ContainerRuntimeEndpoint set the container runtime(s) endpoint(s)
+	// Deprecated
```
Make this `Deprecated: <description>` c.f.: https://github.com/golang/go/wiki/Deprecated

ianvernon commented on pkg/option/config.go:
```diff
@@ -290,6 +292,7 @@ const (
 	// FlannelManageExistingContainers sets if Cilium should install the BPF
 	// programs on already running interfaces created by flannel. Require
 	// Cilium to be running in the hostPID.
+	// Deprecated
```
Make this `Deprecated: <description>` c.f.: https://github.com/golang/go/wiki/Deprecated

ianvernon commented on daemon/endpoint.go:
```diff
@@ -307,7 +307,11 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
 func (h *putEndpointID) Handle(params PutEndpointIDParams) middleware.Responder {
-	log.WithField(logfields.Params, logfields.Repr(params)).Debug("PUT /endpoint/{id} request")
+	if ep := params.Endpoint; ep != nil {
+		log.WithField("endpoint", logfields.Repr(*ep)).Debug("PUT /endpoint/{id} request")
```
Non-blocking nit: Is there ever a case where we will care about the rest of the contents of `params` if `params.Endpoint` is not nil? I thought something like `%+v` would show all fields within `params` and not just a subset of them?

aanm replied:
The `ID`?

ianvernon commented on test/runtime/connectivity.go:
```diff
@@ -21,7 +35,6 @@ func runOnNetNextOnly(f func()) func() {
 	return func() {}
 }
 
-var _ = Describe("RuntimeConnectivityInVethModeTest", runtimeConnectivityTest("veth"))
```
I don't fully understand this commit message - doesn't d391e52ee93777a031ec9337aa65b8d01eb2c63d add the way of plumbing the labels through to `cilium-agent`?

aanm replied:
Yes it does but it only works for containers with the network managed by libnetwork, not CNI.

ianvernon commented on plugins/cilium-docker/driver/driver.go:
```diff
@@ -77,19 +82,28 @@ func newLibnetworkRoute(route route.Route) api.StaticRoute {
-	d := &driver{client: c}
+	scopedLog = scopedLog.WithField("dockerHostPath", dockerHostPath)
+	dockerCli, err := dockerCliAPI.NewClientWithOpts(
+		dockerCliAPI.WithVersion("v1.21"),
```
Any specific need for this version? If so, hardcode it into a constant and state why it is used for self-documenting code.

aanm replied:
It's the same being used today. https://github.com/cilium/cilium/blob/cf7e154e73f4ca76faa28357753b7af6dfb282f8/pkg/workloads/docker.go#L103

ianvernon commented on pkg/endpoint/api.go:
```diff
@@ -509,8 +516,20 @@ func (e *Endpoint) ProcessChangeRequest(newEp *Endpoint, validPatchTransitionSta
+	rev := e.replaceIdentityLabels(newEp.OpLabels.IdentityLabels())
+	if rev != 0 {
+		// Run as a go routine since the runIdentityResolver needs to get the lock
+		go e.runIdentityResolver(context.Background(), rev, false)
```
`context.Background()` --> `e.aliveCtx`?

ianvernon reviewed (commented):
Looks great overall, but I do have a few nits / questions.

ianvernon reviewed (approved):
🎉 🎉 🎉
